# Homework

## Question 1

- Install uv
- What's the version of uv you installed?
- Use --version to find out

In [2]:
!uv --version

uv 0.8.18 (c4c47814a 2025-09-17)


A/ `uv 0.8.18 (c4c47814a 2025-09-17)`

In [3]:
!uv remove scikit-learn

Resolved 47 packages in 509ms                                        
Uninstalled 4 packages in 318ms
 - joblib==1.5.2
 - scikit-learn==1.7.2
 - scipy==1.16.2
 - threadpoolctl==3.6.0


## Question 2

- Use uv to install Scikit-Learn version 1.6.1
- What's the first hash for Scikit-Learn you get in the lock file?
- Include the entire string starting with sha256:, don't include quotes

In [4]:
!uv add 'scikit-learn==1.6.1'

Resolved 51 packages in 416ms                                        
⠙ Preparing packages... (0/1)                                                   
⠙ Preparing packages... (0/1)--------------     0 B/10.58 MiB           
⠹ Preparing packages... (0/1)-------------- 14.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 14.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 30.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 46.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 62.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 78.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 94.90 KiB/10.58 MiB         
⠹ Preparing packages... (0/1)-------------- 110.90 KiB/10.58 MiB        
⠹ Preparing packages... (0/1)-------------- 126.90 KiB/10.58 MiB        
⠹ Preparing packages... (0/1)-------------- 142.90 KiB/10.58 MiB        
⠹ Preparing packages... (0/1)-------------- 15

A/ `sha256:b4fc2525eca2c69a59260f583c56a7557c6ccdf8deafdba6e060f94c1c59738e`

## Models

We have prepared a pipeline with a dictionary vectorizer and a model.

It was trained (roughly) using this code:

```python
categorical = ['lead_source']
numeric = ['number_of_courses_viewed', 'annual_income']

df[categorical] = df[categorical].fillna('NA')
df[numeric] = df[numeric].fillna(0)

train_dict = df[categorical + numeric].to_dict(orient='records')

pipeline = make_pipeline(
    DictVectorizer(),
    LogisticRegression(solver='liblinear')
)

pipeline.fit(train_dict, y_train)
```

> Note: You don't need to train the model. This code is just for your reference.

And then saved with Picke. Download it [here](https://github.com/DataTalksClub/machine-learning-zoomcamp/tree/master/cohorts/2025/05-deployment/pipeline_v1.bin)

With `wget`:


In [5]:
!wget https://github.com/DataTalksClub/machine-learning-zoomcamp/raw/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin

--2025-10-27 12:13:45--  https://github.com/DataTalksClub/machine-learning-zoomcamp/raw/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin [following]
--2025-10-27 12:13:45--  https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8001::154, 2606:50c0:8000::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1300 (1.3K) [application/octet-stream]
Saving to: ‘pipeline_v1.bin’


## Question 3

Let's use the model!

Write a script for loading the pipeline with pickle
Score this record:

```json
{
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}
```
What's the probability that this lead will convert?

- 0.333
- 0.533
- 0.733
- 0.933

If you're getting errors when unpickling the files, check their checksum:

```bash
$ md5sum pipeline_v1.bin
7d17d2e4dfbaf1e408e1a62e6e880d49 *pipeline_v1.bin
```

In [6]:
!md5sum pipeline_v1.bin

7d17d2e4dfbaf1e408e1a62e6e880d49  pipeline_v1.bin


In [9]:
import pickle

with open('pipeline_v1.bin', 'rb') as f:
    dv, model = pickle.load(f)


record = {
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

X = dv.transform([record])

probability = round(model.predict_proba(X)[0, 1], 4)

print("Probability", probability)

Probability 0.5336


A/ `0.5336`

## Question 4

Now let's serve this model as a web service

- Install FastAPI
- Write FastAPI code for serving the model
- Now score this client using requests:

```python
url = "YOUR_URL"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()
```

What's the probability that this client will get a subscription?

- 0.334
- 0.534
- 0.734
- 0.934

In [10]:
!uv add fastapi uvicorn

Resolved 65 packages in 824ms                                        
⠙ Preparing packages... (0/5)                                                   
⠙ Preparing packages... (0/5)--------------     0 B/105.72 KiB          
⠙ Preparing packages... (0/5)-------------- 14.88 KiB/105.72 KiB        
⠙ Preparing packages... (0/5)-------------- 30.88 KiB/105.72 KiB        
⠙ Preparing packages... (0/5)-------------- 46.88 KiB/105.72 KiB        
⠙ Preparing packages... (0/5)2m------------ 62.88 KiB/105.72 KiB        
⠙ Preparing packages... (0/5)---------- 78.88 KiB/105.72 KiB        
⠙ Preparing packages... (0/5)---------- 78.88 KiB/105.72 KiB        
fastapi              ------------------------------ 78.88 KiB/105.72 KiB
⠙ Preparing packages... (0/5)--------------     0 B/451.59 KiB          
fastapi              ------------------------------ 78.88 KiB/105.72 KiB
⠙ Preparing packages... (0/5)--------------     0 B/451.59 KiB          
fastapi              ------------------------------ 78

In [ ]:
!uv add requests

Resolved 69 packages in 441ms                                        
Installed 4 packages in 13ms.4.4                                 
 + certifi==2025.10.5
 + charset-normalizer==3.4.4
 + requests==2.32.5
 + urllib3==2.5.0


In [13]:
import requests

In [16]:
url = "http://localhost:9696/predict"

client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}

response = requests.post(url, json=client)
predictions = response.json()

print(predictions)

{'subscribe_probability': 0.5340417283801275, 'subscribe': True}


A/ `0.534`

## Docker

For these questions, we prepared a base image: `agrigorev/zoomcamp-model:2025`.

This image is based on `3.13.5-slim-bookworm` and has a pipeline with logistic regression (a different one) as well a dictionary vectorizer inside.

This is how the Dockerfile for this image looks like:

```Docker
FROM python:3.13.5-slim-bookworm
WORKDIR /code
COPY pipeline_v2.bin .
```

We already built it and then pushed it to [agrigorev/zoomcamp-model:2025](https://hub.docker.com/r/agrigorev/zoomcamp-model).

> Note: You don't need to build this docker image, it's just for your reference.

## Question 5

Download the base image `agrigorev/zoomcamp-model:2025`. You can easily make it by using docker pull command.

So what's the size of this base image?

- 45 MB
- 121 MB
- 245 MB
- 330 MB

You can get this information when running docker images - it'll be in the "SIZE" column.

In [18]:
!docker image ls | grep agrigorev

agrigorev/zoomcamp-model                        2025              4a9ecc576ae9   6 days ago      121MB


A/ `121MB`

## Dockerfile


Now create your own Dockerfile based on the image we prepared.

It should start like that:

```Docker
FROM agrigorev/zoomcamp-model:2025
# add your stuff here
```

Now complete it:

- Install all the dependencies from pyproject.toml
- Copy your FastAPI script
- Run it with uvicorn

After that, you can build your docker image.


# Question 6

Let's run your docker container!

After running it, score this client once again:

```python
url = "YOUR_URL"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()
```

What's the probability that this lead will convert?

- 0.39
- 0.59
- 0.79
- 0.99


In [19]:
url = "http://localhost:9696/predict"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()

{'subscribe_probability': 0.9933071490756734, 'subscribe': True}

A/ `0.99`